PyRDF/Node.py

```diff
@@ -171,3 +176,176 @@ def graph_prune(self):
 
         self.children = children
         return self.is_prunable()
+
+class HeadNode(Node):
+    """
+    The Python equivalent of ROOT C++'s
+    RDataFrame class.
+
+    Attributes:
+        args (list): A list of arguments that were provided to construct
+            the RDataFrame object.
+
+
+    PyRDF's RDataFrame constructor accepts the same arguments as the ROOT's
+    RDataFrame constructor (see
+    `RDataFrame <https://root.cern/doc/master/classROOT_1_1RDataFrame.html>`_)
+
+    In addition, PyRDF allows you to use Python lists in place of C++ vectors
```

Comment:
This is no longer an exclusive feature of PyRDF since new cppyy has a converter Python list -> std::vector. Anyway, we are no longer going to construct RDataFrames like this anymore, but just via the factory functions?

Reply:
Or is this intended to be used by the backends?

Reply:
I remember we discussed this last year and decided to keep it for backward compatibility. Since the plan now is to integrate in ROOT, anything that is the default in master today can be taken for granted I believe. In this case I can remove this logic from PyRDF.
Does this sound good ?

Reply:
Yes I would remove this.

Reply:
Who is intended to use this constructor on the other hand? The backends?

Reply:
Whomever will create any PyRDF.WhateverBackendDataFrame . Also the backends in the sense that `mapper` function creates ROOT.RDataFrame objects on the nodes, so this constructor will also be involved there

Reply:
Ok so it's the backends, never the user code directly (PyRDF.WhateverBackendDF is also backend).

Reply:
Another thing here, there's a test that checks this functionality specifically
```python
def test_two_args(self):
    """Constructor with list of input files"""
    rdf_2_files = ["file1.root", "file2.root"]

    # Convert RDF files list to ROOT CPP vector
    reqd_vec = ROOT.std.vector('string')()
    for elem in rdf_2_files:
        reqd_vec.push_back(elem)

    # RDataFrame constructor with 2nd argument as string
    hn_1 = Node.HeadNode("treename", "file.root")

    # RDataFrame constructor with 2nd argument as Python list
    hn_2 = Node.HeadNode("treename", rdf_2_files)

    # RDataFrame constructor with 2nd argument as ROOT CPP Vector
    hn_3 = Node.HeadNode("treename", reqd_vec)

    self.assertArgs(hn_1.args, ["treename", "file.root"])
    self.assertArgs(hn_2.args, ["treename", reqd_vec])
    self.assertArgs(hn_3.args, ["treename", reqd_vec])
```

And this has made me wonder, is there a constructor in PyROOT that accepts `ROOT.std.vector("string")` as filelist? In that case I should maintain this in PyRDF too

Reply:
PyROOT has now a converter for Python list -> std::vector thanks to cppyy. This means you don't need to maintain this conversion in PyRDF anymore, it is provided by the bindings.

Reply:
Yep but I still saw some errors in the tests so I was confused. I might have found a bug in the conversion of std::string with TChain
```python
>>> import ROOT
>>> v = ROOT.std.vector("string")()
>>> for name in ["2clusters.root","4clusters.root"]:
...     v.push_back(name)
... 
>>> df = ROOT.RDataFrame("myTree",v)
>>> ch = ROOT.TChain("myTree")
>>> for name in v:
...     ch.Add(name)
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
TypeError: none of the 2 overloaded methods succeeded. Full details:
  int TChain::Add(TChain* chain) =>
    TypeError: could not convert argument 1
  int TChain::Add(const char* name, Long64_t nentries = TTree::kMaxEntries) =>
    TypeError: could not convert argument 1 (bad argument type for built-in operation)
>>> for name in v:
...     print(type(name))
...     print(name)
... 
<class cppyy.gbl.std.string at 0x55ef1a9f4d50>
2clusters.root
<class cppyy.gbl.std.string at 0x55ef1a9f4d50>
4clusters.root
```

Reply:
The problem here is that, in new cppyy, when you iterate over an `std::vector<std::string>`, what you get back is `std::strings` - `name` is an `std::string`. But `TChain:Add` expects a `const char*`. The solution is not to use `std::vector` anymore in the test.

Reply:
Still there exists the possibility of constructing an RDataFrame with an std vector in python.

```python
>>> v = ROOT.std.vector("string")()
>>> for name in ["one.root","two.root"]:
...     v.push_back(name)
... 
>>> df = ROOT.RDataFrame("myTree", v)
>>> df
<cppyy.gbl.ROOT.RDataFrame object at 0x562ada82e910>
```
So that needs to be supported in PyRDF too. And since we make use of TChain, I believe the solution is to keep the conversion to a python string when constructing the TChain

Reply:
Yes, it's ok to test the constructor of `RDataFrame` that accepts an `std::vector<std::string>`, just don't use what you get out of iterating that vector in `TChain::Add` (or do the conversion to Python str as you said).